compress CAS blobs on upload

rmaz opened the pull request:
Compress CAS blobs on upload and try and decompress on download. If decompression fails, fall back to uncompressed content fetching. This allows for phasing in the compression without requiring a cache flush. The opposite case is not handled, we would have to flush the cache to disable compression again.

Commit:
compress CAS blobs on upload

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber-common/bazel?pullRequest=23) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber-common/bazel?pullRequest=23) before we can accept your contribution.<br/><hr/>**Richard Howell** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber-common/bazel?pullRequest=23) it.</sub>

rmaz commented on src/main/java/com/google/devtools/build/lib/remote/http/HttpUploadHandler.java:
```diff
@@ -138,8 +143,41 @@ private HttpRequest buildRequest(String path, String host, long contentLength) {
+  private HttpChunkedInput buildBody(UploadCommand msg) throws IOException {
+    if (msg.casUpload()) {
+      // we first have to compress to a byte buffer
+      // as we need to know the content length for the header
```
This part is extremely ugly, I was hoping we could get away with just streaming the compression to the network, but bazel-remote requires the Content-Length header (even though with chunked input my understanding was you should not have this present). Compressing buffers in memory in Java is super clunky.

AlessandroPatti commented:
Can we put the compression behind a flag?

AlessandroPatti commented on src/main/java/com/google/devtools/build/lib/remote/http/HttpUploadHandler.java:
```diff
@@ -138,8 +143,41 @@ private HttpRequest buildRequest(String path, String host, long contentLength) {
-  private HttpChunkedInput buildBody(UploadCommand msg) {
-    return new HttpChunkedInput(new ChunkedStream(msg.data()));
+  private HttpChunkedInput buildBody(UploadCommand msg) throws IOException {
+    if (msg.casUpload()) {
```
How long does it take to do this? Might this trigger the idle timeout?

rmaz replied:
Good question, will investigate. Has the timer started already at this point?

rmaz replied:
It does, yes, although after a much longer time than expected. I added a 10s sleep and set the remote timeout to 1s, this fired one timeout for a static archive after compression but before upload. Setting this to a 2s sleep did not trigger. Need to investigate whats going on here.

rmaz commented:
> Can we put the compression behind a flag?

Can do, yes. The issue would be that once we have this enabled you can't go back again, the cache is poisioned, so you would have to flush everything to get it working again.

Commit:
compress in cache client instead

Commit:
add flag

Commit:
close the compression stream

AlessandroPatti commented on src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:
```diff
@@ -740,8 +751,20 @@ public void close() {
+    File compressedUpload = null;
     if (storedBlobs.putIfAbsent((casUpload ? CAS_PREFIX : AC_PREFIX) + key, true) == null) {
       try {
+        if (casUpload && this.compressCasUploads) {
+          // first compress to a temporary file before creating an upload channel
+          compressedUpload = createCompressedUpload(key, in);
+          in.close();
+          in = new FileInputStream(compressedUpload);
+          InputStream wrappedCompressedIn = new FilterInputStream(in) {
+            @Override
+            public void close() {}
+          };
+          upload = new UploadCommand(uri, true, key, wrappedCompressedIn, compressedUpload.length());
+        }
```
Can we move this up and wrap only once? Something on the lines of
```java
InputStream input;
if (casUpload && this.compressCasUploads) {
    compressedUpload = createCompressedUpload(key, in);
    input = new FileInputStream(compressedUpload);
else {
    input = in;
}
InputStream wrappedIn =
    new FilterInputStream(input) { 
```

rmaz replied:
👍

AlessandroPatti commented on src/main/java/com/google/devtools/build/lib/remote/http/HttpDownloadHandler.java:
```diff
@@ -105,8 +116,18 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpObject msg) throws Ex
       int readableBytes = content.readableBytes();
-      content.readBytes(out, readableBytes);
+      int readerIndex = content.readerIndex();
+      try {
+        content.readBytes(out, readableBytes);
+      } catch (ZipException ex) {
+        // not a compressed response, fall back to uncompressed
+        logger.log(Level.WARNING, "Decompression failed: " + ex.toString());
+        content.readerIndex(readerIndex);
+        out = rawOut;
+        content.readBytes(out, readableBytes);
+      }
```
Can we handle this in `HttpCacheClient` based on the option value rather than passing it down to the handler?

rmaz replied:
We can wrap the output stream there, yes. May be a bit fiddly to get the fallback working though. Will take a look.

AlessandroPatti commented on src/main/java/com/google/devtools/build/lib/remote/options/RemoteOptions.java:
```diff
@@ -441,6 +441,15 @@ public RemoteOutputsStrategyConverter() {
+      defaultValue =  "false",
+      documentationCategory = OptionDocumentationCategory.REMOTE,
+      effectTags = {OptionEffectTag.EXECUTION},
+      help = "If set to true, Bazel will compress CAS blobs before upload and "
+        + "try to decompress them on download.")
```
Please recall that this is only works with HTTP in the help message.

AlessandroPatti commented:
Needs rebase on 3.2.0

Commit:
refactor for single input wrapping

Commit:
move decompression to HttpCacheClient

Commit:
update help message

rmaz commented:
Rebased and updated. No longer handles fallback on zip decompression errors, but probably makes more sense this way and code is simpler. We're going to need a way to flush the cache when switching this on though.

AlessandroPatti commented on src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:
```diff
@@ -774,6 +808,19 @@ public void close() {
+    File compressedOutput = File.createTempFile(name, ".gz");
+    compressedOutput.deleteOnExit();
+    OutputStream out = new DeflaterOutputStream(new FileOutputStream(compressedOutput));
+    byte[] buffer = new byte[131072];
+    int len;
+    while ((len = input.read(buffer)) > 0) {
+      out.write(buffer, 0, len);
+    }
```
Nit: can be simplified with `ByteStreams.copy`

rmaz replied:
Doesn't say what buffer size it uses, prefer doing it this way.

AlessandroPatti commented on src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:
```diff
@@ -774,6 +808,19 @@ public void close() {
 
+  private File createCompressedUpload(String name, InputStream input) throws IOException {
+    File compressedOutput = File.createTempFile(name, ".gz");
+    compressedOutput.deleteOnExit();
```
Nit: this might not have the desired effect as bazel runs with a persistent server

rmaz replied:
Agreed, this is just a cautious fallback in case of JVM shutdown before the request completes.

AlessandroPatti commented:
LGTM
